src/main/kotlin/com/exactpro/th2/ws/client/Main.kt

```diff
@@ -0,0 +1,191 @@
+) {
+    val connectionId = ConnectionID.newBuilder().setSessionAlias(settings.sessionAlias).build()
+
+    val rootEventId = eventRouter.storeEvent(Event.start().apply {
+        name("WS client '${settings.sessionAlias}' ${Instant.now()}")
+        type("Microservice")
+    }).id
+
+    settings.handlerSettings.runCatching(handler::init).onFailure {
+        LOGGER.error(it) { "Failed to init request handler" }
+        eventRouter.storeEvent(rootEventId, "Failed to init request handler", "Error", it)
+        throw it
+    }
+
+    val incomingSequence = createSequence()
+    val outgoingSequence = createSequence()
+
+    //TODO: add batching
+    val onMessage = { message: ByteArray, _: Boolean, direction: Direction ->
+        val sequence = if (direction == Direction.FIRST) incomingSequence else outgoingSequence
+        val attribute = if (direction == Direction.FIRST) QueueAttribute.FIRST else QueueAttribute.SECOND
+        messageRouter.send(message.toBatch(connectionId, direction, sequence()), attribute.toString())
+    }
+
+    val onEvent = { cause: Throwable?, message: () -> String ->
+        val type = if (cause != null) "Error" else "Info"
+        eventRouter.storeEvent(rootEventId, message(), type, cause)
+    }
+
+    val client = WebSocketClient(
+        URI(settings.uri),
+        handler,
+        onMessage,
+        onEvent
+    ).apply { registerResource("client", ::close) }
+
+    val listener = MessageListener<MessageGroupBatch> { _, message ->
+        message.groupsList.forEach { group ->
+            group.runCatching {
+                require(messagesCount == 1) { "Message group contains more than 1 message" }
+                val message = messagesList[0]
+                require(message.hasRawMessage()) { "Message in the group is not a raw message" }
+                val bytes = message.rawMessage.body.toByteArray()
+                if (settings.frameType == TEXT) client.sendText(bytes.toString(UTF_8)) else client.sendBinary(bytes)
+            }.recoverCatching {
+                LOGGER.error(it) { "Failed to handle message group: ${group.toPrettyString()}" }
+                eventRouter.storeEvent(rootEventId, "Failed to handle message group: ${group.toPrettyString()}", "Error", it)
+            }
+        }
+    }
+
+    runCatching {
+        checkNotNull(messageRouter.subscribe(listener, INPUT_QUEUE_ATTRIBUTE))
+    }.onSuccess { monitor ->
+        registerResource("raw-monitor", monitor::unsubscribe)
+    }.onFailure {
+        throw IllegalStateException("Failed to subscribe to input queue", it)
+    }
+
+    LOGGER.info { "Successfully started" }
+    @Suppress("ControlFlowWithEmptyBody")
+    while (client.isRunning);
+    LOGGER.info { "Finished running" }
+}
+
+data class Settings(
+    val uri: String,
+    val frameType: FrameType = TEXT,
+    val sessionAlias: String,
+    val handlerSettings: IHandlerSettings? = null
+) {
+    enum class FrameType {
+        TEXT,
+        BINARY
```

Comment:
Maybe add sned(client: WebSocketClient, bytes: ByteArray) abstract method instead of using if blocks

Reply:
You mean that we should initialize `IClient` instance with frame type, remove `sendText`, `sendBinary` methods from this interface and create a `send(client: IClient, bytes: ByteArray)` method which would incapsulate logic of selecting a proper send method?

Reply:
I mean we can replace `if (settings.frameType == TEXT) client.sendText(bytes.toString(UTF_8)) else client.sendBinary(bytes)` to call of a method from FrameType, for example settings.frameType.send(client, bytes)